source/common/thread_local/thread_local_impl.cc

```diff
@@ -136,15 +139,43 @@ void InstanceImpl::removeSlot(uint32_t slot) {
              free_slot_indexes_.end(),
          fmt::format("slot index {} already in free slot set!", slot));
   free_slot_indexes_.push_back(slot);
-  runOnAllThreads([slot]() -> void {
+
+  // Find out how many dispatchers have associated running thread and are able to process
+  // events. During normal operation, all dispatchers should be in Running state, but
+  // during unit and integration tests, dispatchers are sometimes allocated without
+  // worker thread. Events posted to such dispatchers are queued but never executed.
+  // The following logic determines how many threads are in Running state.
+  // An event to deallocate a slot is posted to each thread and main thread blocks until
+  // all Running threads report that they have finished deallocating the slot.
+
+  // Include main thread.
+  auto running_workers = 1;
+  std::for_each(registered_threads_.begin(), registered_threads_.end(),
+                [&running_workers](std::reference_wrapper<Event::Dispatcher>& dispatcher) {
+                  if (dispatcher.get().isRunning()) {
+                    running_workers++;
+                  }
+                });
+
+  std::shared_ptr<absl::BlockingCounter> counter =
+      std::make_shared<absl::BlockingCounter>(running_workers);
+  std::weak_ptr<absl::BlockingCounter> counter_weak = counter;
+  runOnAllThreads([slot, counter_weak]() -> void {
     // This runs on each thread and clears the slot, making it available for a new allocations.
     // This is safe even if a new allocation comes in, because everything happens with post() and
     // will be sequenced after this removal. It is also safe if there are callbacks pending on
     // other threads because they will run first.
     if (slot < thread_local_data_.data_.size()) {
       thread_local_data_.data_[slot] = nullptr;
     }
+    if (auto counter = counter_weak.lock()) {
+      counter->DecrementCount();
+    }
   });
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.thread_local_synchronize_slots_deletion")) {
+    counter->Wait();
```

Comment:
ugh so sorry, I think I gave you poor advice by not actually understanding the underlying problem.

Ok here's my new thought, but I'd appreciate a chime in from @mattklein123 in case there's some clever workaround because my last suggestion wasn't great :-/

Tldr for Matt please correct if not accurate @cpakulski  from the linked issue and first attempt, each DFP filter registers with the global CM for cluster updates.  when the DFP filter goes away, there's a race where the ClusterUpdateCallbacksHandlePtr is deleted but the the main thread is still sending updates.

I now think the correct thing to do is add a big warning note about how thread local slots will be deleted and potential race conditions.  Then make ClusterUpdateCallbacksHandlePtr have a release() function which makes up-calls stop immediately, then either make it a shared pointer or hand ownership to a post to the main thread dispatcher so it's not deleted until after the main thread stops referencing it.
/wait-any

Reply:
> ugh so sorry, I think I gave you poor advice by not actually understanding the underlying problem.

@alyssawilk, you actually gave the correct advice to try shared pointer: https://github.com/envoyproxy/envoy/pull/32692#pullrequestreview-1923107947. 
That would however require some restructuring of DFP filter and maybe some other filters which register callbacks. I thought that before trying that, maybe the problem can be fixed at thread local level. My next step is to try intermediate shared pointer, which can be used to wrap callbacks embedded in DFP filter. If it works, it will also be generic solution and will not require DFP filter changes. I am working on that. If it does not work or is ugly, I will try to restructure DFP filter. 

> Tldr for Matt please correct if not accurate @cpakulski from the linked issue and first attempt, each DFP filter registers with the global CM for cluster updates. when the DFP filter goes away, there's a race where the ClusterUpdateCallbacksHandlePtr is deleted but the the main thread is still sending updates.

That is close description, except one small detail:

each DFP filter registers with the global CM for cluster updates. when the DFP filter goes away, there's a race where one _of the worker threads has not-finished yet (scheduled or is currently executing) job which references callbacks which were deleted by DFP on main thread_.

Reply:
Sorry I would need to sit down with the code and look at it all to have good advice. I can do that if needed. In general I agree with @kyessenov that in the past we have extended the lifetimes of things as needed and/or made sure that destruction happens on the main thread for cases like this.

Reply:
oh phew.  yeah I think having a cross-thread lock is going to be more problematic than shared pointer, which is also inline with Matt's advice about extending lifetime so I think we want to go that route.